eclass/zig-toolchain.eclass

```diff
@@ -0,0 +1,325 @@
+	esac
+
+	case ${c_target_prefix} in
+		arm64)	;;
+		arm*)
+			local is_softfloat=$(CTARGET=${c_target} tc-tuple-is-softfloat)
+			case ${is_softfloat} in
+				only | yes) features+="+soft_float";;
+				softfp | no) features+="-soft_float";;
+				*) die "${FUNCNAME[0]}: tc-tuple-is-softfloat returned unexpected value"
+			esac
+		;;
+	esac
+
+	echo ${base_cpu}${features}
+}
+
+# @FUNCTION: zig-toolchain_find_installation
+# @DESCRIPTION:
+# Detects suitable Zig installation and sets ZIG_VER and ZIG_EXE variables.
+#
+# See ZIG_EXE and ZIG_VER descriptions for more information.
+zig-toolchain_find_installation() {
+	# Adapted from https://github.com/gentoo/gentoo/pull/28986
+	# Many thanks to Florian Schmaus (Flowdalic)!
+
+	[[ -n ${ZIG_SLOT} ]] || die "${FUNCNAME[0]}: ZIG_SLOT must be set"
+	if ver_test "${ZIG_SLOT}" -lt "0.13"; then
+		die "${ECLASS}: ZIG_SLOT must be >= 0.13, found ${ZIG_SLOT}"
+	fi
+
+	einfo "Searching Zig ${ZIG_SLOT}..."
+
+	local selected_path selected_version
+
+	# Prefer "dev-lang/zig" over "dev-lang/zig-bin"
+	local -a candidates candidates1 candidates2
+	readarray -d '' -t candidates1 < <(find -P "${BROOT}"/usr/bin/ -mindepth 1 -maxdepth 1 -type l -name "zig-bin-*" -print0 2> /dev/null || echo -n "")
+	readarray -d '' -t candidates2 < <(find -P "${BROOT}"/usr/bin/ -mindepth 1 -maxdepth 1 -type l -name "zig-*" -a '!' -name "zig-bin-*" -print0 2> /dev/null || echo -n "")
+	candidates=( "${candidates1[@]}" "${candidates2[@]}" )
+
+	local candidate_path
+	for candidate_path in "${candidates[@]}"; do
+		local candidate_version="${candidate_path##*-}"
+
+		# Compare with ZIG_SLOT (like 0.13)
+		local candidate_slot=$(ver_cut 1-2 ${candidate_version})
+		if ver_test "${candidate_slot}" -ne ${ZIG_SLOT}; then
+			# Candidate does not satisfy ZIG_SLOT condition.
+			continue
+		fi
+
+		# Compare with previous version (like 0.13.0 and 0.13.1, we prefer the second one if possible)
+		if [[ -n "${selected_path}" ]]; then
+			local selected_version="${selected_path##*-}"
+			if ver_test ${candidate_version} -lt ${selected_version}; then
+				# Previously chosen version is higher, keep it.
+				continue
+			fi
+		fi
+
+		selected_path="${candidate_path}"
+	done
+
+	if [[ -z "${selected_path}" ]]; then
+		die "Could not find (suitable) Zig installation in \"${BROOT}/usr/bin/\""
+	fi
+
+	export ZIG_EXE="${selected_path}"
+}
+
+# @FUNCTION: zig-toolchain_populate_env_vars
+# @DESCRIPTION:
+# Populates ZIG_TARGET, ZIG_CPU, ZIG_EXE and ZIG_VER environment
+# variables with detected values, or, if user set them already,
+# leaves as-is.
+zig-toolchain_populate_env_vars() {
+	# Should be first because it sets ZIG_VER which might be used in the future
+	# when setting ZIG_TARGET and ZIG_CPU variables for incompatible versions.
+	if [[ -z "${ZIG_EXE}" ]]; then
+		zig-toolchain_find_installation
+	fi
+	if [[ -z ${ZIG_VER} ]]; then
+		local selected_ver=$("${ZIG_EXE}" version || die "Failed to get version from \"${ZIG_EXE}\"")
```

Comment:
Why do we need to do this? Why can't we go off best_version/has_version for whatever slot we found from iteration?

Reply:
1) Because for live slot result would be different, "0.14.0-dev.smth+123" vs 9999 and first one is more useful to make version-conditional logic
2) To check that `zig` at least works to print version (it pulls `checkLibcxx` and smth similar inside IIRC)

Reply:
It doesn't seem to be used anywhere right now, unless `zig` itself uses `ZIG_VER`. It seems odd to me that we would find it useful -- for 9999, we would want to adapt immediately and unconditionally (not based on some dev version) as we do for other ecosystems. And for the other case, I think we could really do with seeing an example where it's needed first.

But 2. is a fair enough point, I suppose. But if that's the only reason remaining, it should be branded as a sanity check instead.

Reply:
To give another angle: this kind of logic, once in, tends to stay there forever. I'd much rather us revisit it when we have a concrete case to analyse, rather than having something which may bitrot.

Reply:
I have used it succesfully to differentiate 2 Zig 999 versions where I set EGIT_COMMIT one as 0.12-x (or maybe older) branch with 0.12.1 pre-release and main branch. In that case logic "< 0.13" worked as expected. This is rare case handcrafted by myself

> But 2. is a fair enough point, I suppose. But if that's the only reason remaining, it should be branded as a sanity check instead.
Like this?
```bash
export ZIG_VER=${selected_ver}
# sanity check
zig version > /dev/null || die "Zig version check failed!"
```

Reply:
> I have used it succesfully to differentiate 2 Zig 999 versions where I set EGIT_COMMIT one as 0.12-x (or maybe older) branch with 0.12.1 pre-release and main branch. In that case logic "< 0.13" worked as expected. This is rare 
case handcrafted by myself

But why did you need it, and why couldn't you hand-roll that in a specific ebuild in the event you did need it?

Reply:
> Like this?

Right.

Reply:
3767c578f1d7f21637fc19e6d19e407cd2bbf423